autoendpoint/src/routers/stub/router.rs

```diff
@@ -0,0 +1,96 @@
+use super::client::StubClient;
+use super::error::StubError;
+use super::settings::{StubServerSettings, StubSettings};
+use crate::error::ApiResult;
+use crate::extractors::notification::Notification;
+use crate::extractors::router_data_input::RouterDataInput;
+use crate::routers::{Router, RouterError, RouterResponse};
+use async_trait::async_trait;
+use serde_json::Value;
+use std::collections::HashMap;
+
+/// Firebase Cloud Messaging router
+pub struct StubRouter {
+    settings: StubSettings,
+    server_settings: StubServerSettings,
+    /// A map from application ID to an authenticated FCM client
+    clients: HashMap<String, StubClient>,
+}
+
+impl StubRouter {
+    /// Create a new `StubRouter`
+    pub fn new(settings: StubSettings) -> Result<Self, StubError> {
+        let server_settings =
+            serde_json::from_str::<StubServerSettings>(&settings.server_credentials)
+                .unwrap_or_default();
+        let clients = Self::create_clients(&server_settings);
+        Ok(Self {
+            settings,
+            server_settings,
+            clients,
+        })
+    }
+
+    /// Create Test clients for each application
+    fn create_clients(server_settings: &StubServerSettings) -> HashMap<String, StubClient> {
+        let mut clients = HashMap::new();
+
+        clients.insert("success".to_owned(), StubClient::success());
+        clients.insert(
+            "error".to_owned(),
+            StubClient::error(StubError::General(server_settings.error.clone())),
+        );
+        clients
+    }
+}
+
+#[async_trait(?Send)]
+impl Router for StubRouter {
+    fn register(
+        &self,
+        router_data_input: &RouterDataInput,
+        app_id: &str,
+    ) -> Result<HashMap<String, Value>, RouterError> {
+        if !self.clients.contains_key(app_id) {
+            return Err(
+                StubError::General(format!("Unknown App ID: {}", app_id.to_owned())).into(),
+            );
+        }
+
+        let mut router_data = HashMap::new();
+        router_data.insert(
+            "token".to_string(),
+            serde_json::to_value(&router_data_input.token).unwrap(),
+        );
+        router_data.insert("app_id".to_string(), serde_json::to_value(app_id).unwrap());
```

Comment:
A minor nit, but there's a mix of `to_string` and `to_owned` in this PR, let's just stick with one
```suggestion
            "token".to_owned(),
            serde_json::to_value(&router_data_input.token).unwrap(),
        );
        router_data.insert("app_id".to_owned(), serde_json::to_value(app_id).unwrap());
```